### PR TITLE
Await approval request in transaction controller

### PR DESCRIPTION
**Pull request by matthewwalsh0**
## Explanation

Refactor the transaction controller to decouple the public methods from the concept of confirmation by instead awaiting the approval request promise and then performing the approve or cancel logic, all in a single method execution.

At a technical level, this includes:

- Updating the `newUnapprovedTransaction` and `addUnapprovedTransaction` methods to await approval and continue with the post-approval logic.
- Changing the `approveTransaction` and `cancelTransaction` controller methods to private.
- Adding an optional `requiresApproval` property to `addUnapprovedTransaction` to skip approval.
- Updating the `updateAndApproveTx`, `cancelTx`, and `cancelTxs` actions to resolve or reject the approval using the `ApprovalController`.
- Updating the swaps reducer to add the transaction and skip approval in a single method call.
- Moving swaps specific transaction logic after creating the transaction but before automatic approval, into the transaction controller.

## Manual Testing Steps

Verify all transaction related flows including:

- Transfers
- Contract Operations
- Swaps

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

**Comment by Taniel13** [comment]
> ## Explanation
> 
> Refactor the transaction controller to decouple the public methods from the concept of confirmation by instead awaiting the approval request promise and then performing the approve or cancel logic, all in a single method execution.
> 
> At a technical level, this includes:
> 
> - Updating the `newUnapprovedTransaction` and `addUnapprovedTransaction` methods to await approval and continue with the post-approval logic.
> - Adding an optional `requiresApproval` property to `addUnapprovedTransaction` to skip approval.
> - Updating the `updateAndApproveTx`, `cancelTx`, and `cancelTxs` actions to resolve or reject the approval using the `ApprovalController`.
> - Updating the swaps reducer to add the transaction and skip approval in a single method call.
> - Moving swaps specific transaction logic after creating the transaction but before automatic approval, into the transaction controller.
> 
> ## Manual Testing Steps
> 
> Verify all transaction related flows including:
> 
> - Transfers
> - Contract Operations
> - Swaps
> 
> ## Pre-merge author checklist
> 
> - [x] I've clearly explained:
>   - [x] What problem this PR is solving
>   - [x] How this problem was solved
>   - [x] How reviewers can test my changes
> - [x] Sufficient automated test coverage has been added
> 
> ## Pre-merge reviewer checklist
> 
> - [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
> - [ ] PR is linked to the appropriate GitHub issue
> - [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone
> 
> If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.
> 
> In this case, a QA Engineer approval will be be required.
>

File: app/scripts/controllers/transactions/index.js
Comment (bergarces) [inline app/scripts/controllers/transactions/index.js:1691]: Are we emitting this event only for the unit tests?

If so, we might want to think on how to refactor them, even if it's in a future PR.

File: app/scripts/controllers/transactions/index.js
Comment (bergarces) [inline app/scripts/controllers/transactions/index.js:1763]: Same comment as the previous `newUnapprovedTx` event, but this one is even weirder as `addTransaction` emits a ``${txMeta.id}:unapproved`` with exactly the same data and I cannot see if anyone is listening.

Odd, although we might want to keep them just in case, if they don't seem necessary I'd put a comment for future maintainers.

File: app/scripts/controllers/transactions/index.js
Comment (bergarces) [inline app/scripts/controllers/transactions/index.js:1766]: Totally optional and subjective, but I think it'd be clearer for anyone else reviewing if this whole block of code that comes from `swaps.js` UI logic was moved into a private method, guarded here by just the transaction type. Something like:

```
if (transactionType === TransactionType.swap) {
  this._whateverNameGivenToTheSwapSpecificWork(...)
}
```

But maybe this comment is good enough for any reviewer to see where it comes from and why it's here.

File: app/scripts/controllers/transactions/index.js
Comment (bergarces) [inline app/scripts/controllers/transactions/index.js:743]: From what I can see, `_waitForTransactionFinished` resolves when a `${id}:finished` event is fired.

However, that event is only fired when the status is changed to `submitted`, `rejected` or `failed`.

Since we are just awaiting if the initial status is `unapproved`, is there any chance that we could be missing out on awaiting for the tx from any of the other states (`approved`, `signed`, `dropped`, `confirmed`, `pending`). It might be the case that this is impossible, in which case we are good, but just mentioning it so that it's not overlooked.

File: app/scripts/controllers/transactions/index.js
Comment (bergarces) [inline app/scripts/controllers/transactions/index.js:2768]: This one returns void.

File: app/scripts/controllers/transactions/index.js
Comment (bergarces) [inline app/scripts/controllers/transactions/index.js:1209]: This is a weird one but... now that this is not called externally via `TransactionController` instance, but only via `PendingTransactionTracker` via passing the method in the constructor... should it be private?

File: app/scripts/controllers/transactions/index.js
Comment (cryptotavares) [inline app/scripts/controllers/transactions/index.js:371]: This is a farfetched case, but if the `_requestTransactionApproval` ends up throwing after setting the transaction status to approved but before setting it to submitted, then we will end up with a pending Promise (`finishedPromise`) that will never resolve and will be blocking the execution flow. Should we or even do we want to handle transactions that were not submitted but were a approved? 🤔 

(I guess that this could already happen before anyway, so any solution should be done on a subsequent PR)

File: app/scripts/controllers/transactions/index.js
Comment (matthewwalsh0) [reply app/scripts/controllers/transactions/index.js]: Didn't really like this as I wrote it. I've removed the event and am instead using a spy on the state manager.

File: app/scripts/controllers/transactions/index.js
Comment (matthewwalsh0) [reply app/scripts/controllers/transactions/index.js]: I've left all the state manager events but removed `newUnapprovedTx` as it was now unused.

File: app/scripts/controllers/transactions/index.js
Comment (matthewwalsh0) [reply app/scripts/controllers/transactions/index.js:1684]: Good idea, done.

File: app/scripts/controllers/transactions/index.js
Comment (matthewwalsh0) [reply app/scripts/controllers/transactions/index.js]: Done.

File: app/scripts/controllers/transactions/index.js
Comment (matthewwalsh0) [reply app/scripts/controllers/transactions/index.js]: I didn't notice `PendingTransactionTracker` was constructed inside this controller, so changed to private to encapsulate everything.

File: ui/store/actions.ts
Comment (matthewwalsh0) [reply ui/store/actions.ts]: Updated.

File: app/scripts/controllers/transactions/index.js
Comment (matthewwalsh0) [reply app/scripts/controllers/transactions/index.js:371]: Good spot, I've fixed the condition inside the error handling in `_requestTransactionApproval` so it will now correctly cancel or fail the transaction if it's in any incomplete status. Meaning any errors here are just for information, and the transaction will always be `submitted`, `rejected`, or `failed` at this point.

**Comment by danjm** [comment]
One issue I notice while reviewing this PR is we now have two different uses of the term "approval" within the transaction controller. This is not a huge deal, and it is a problem that existed before this PR, but the changes in this PR make it more noticeable.

To clarify, we have:
1. the concept of a "swapApproval", which is a transaction that calls the ERC20 `approve` method to grant the swaps contract the ablity to transfer tokens
2. the `ApprovalController` and the approvals it controllers

And so it can lead to some confusion for the reader of the code. For example we have code like this:
```
finalApproveTxMeta = await addUnapprovedTransaction(
          undefined,
          { ...approveTxParams, amount: '0x0' },
          TransactionType.swapApproval,
          {
            requireApproval: false,
```
Where `swapApproval` and `requireApproval` are right next to each other, but the `Approval`s in those two property names have entirely different meanings.

Now, this is not a huge deal, because when you read the transaction controller itself it is clear and well written. But a possible source of confusion. Including in `metamask-controller.js`, where we have the handling of the `newSwapApproval` event, but also lots of references to the `approvalController` and approvals that come from that controller, but no documentation to distinguish the `newSwapApproval` from the other sort of approvals.

We probably shouldn't address this problem in this PR, as it was a pre-existing problem. So I'd suggest a clean up task to follow this PR that renames the "swapApproval" to something else, perhaps "swapERC20Approve"

File: app/scripts/controllers/transactions/index.js
Comment (danjm) [inline app/scripts/controllers/transactions/index.js:1725]: What's the rationale for emitting an event here instead of just calling the swaps controller method here?

Comment (danjm) [reply app/scripts/controllers/transactions/index.js:1725]: Oh, maybe because the transactions controller has to be instantiated before the swaps controller?

File: app/scripts/background.js
Comment (danjm) [inline app/scripts/background.js:678]: Why do we no longer need to wait for `initApprovals()` to call `updateBadge()`?

File: ui/store/actions.ts
Comment (danjm) [inline ui/store/actions.ts:1674]: It took me a little while to figure out how the transaction state would correctly be updated now that we are not calling `cancelTransaction` directly. But I think I get it now:
1. `rejectPendingApproval` is called, with the second parameter being the error created by `ethErrors.provider.userRejectedRequest().serialize()`
2. `this.approvalController.reject` is called with an error constructed from its second parameter
3. Steps 1 and 2 will always be happening while we are awaiting `this._requestApproval` in `_requestTransactionApproval` the transaction controller
4. If the user rejects the request, `this._requestApproval` will throw an error
5. The error handler in `_requestTransactionApproval` will check that the error code matches what was set in step 2, and then call `this._cancelTransaction`

Does that sound right?

Comment (danjm) [reply ui/store/actions.ts:1674]: Also, I notice that we are passing `ethErrors.provider.userRejectedRequest().serialize()` here, but in `rejectPendingApproval` in `metamask-controller.js` we assume that the error parameter is not serialized (`new EthereumRpcError(error.code, error.message, error.data)`).

Does deserialization happen automatically by some part of our ui<>background communication code?

**Review by danjm — CHANGES_REQUESTED** [review]
This looks good!

I have left a few questions and when they are answered I can approve.

File: app/scripts/controllers/transactions/index.js
Comment (matthewwalsh0) [reply app/scripts/controllers/transactions/index.js:1725]: Exactly, a circular dependency. We would need the `SwapsController` to exist first to provide the callback in the constructor, but the `SwapsController` requires the `TransactionController` for callbacks in its constructor.

An event enables us to create the listener after both are already constructed.

File: app/scripts/background.js
Comment (matthewwalsh0) [reply app/scripts/background.js:678]: The previous implementation of `initApprovals` wouldn't resolve until the requests were approved, meaning the badge number would be zero at that point anyway.

Plus I noticed we already listened to the `updateBadge` event from the approval controller below anyway, meaning the badge is already reacting to the creation of the approval requests.

File: ui/store/actions.ts
Comment (matthewwalsh0) [reply ui/store/actions.ts:1674]: Exactly right.

The `serialize` method is a bit misleading as it's not serialising to a string, but rather a simple object with only the public properties remaining, including `code`.

https://github.com/MetaMask/rpc-errors/blob/v4.0.3/src/classes.ts#L44

**Review by danjm — DISMISSED** [review]
Nice work!

**Review by OGPoyraz — APPROVED** [review]

**Review by vinistevam — APPROVED** [review]

**Review by bergarces — APPROVED** [review]

**Pull request merged**
